server/data/data-api/src/test/java/org/apache/james/rrt/lib/CanSendFromContract.java

```diff
@@ -93,4 +93,46 @@ default void assertUserCanSendFromShouldBeFalseWhenWhenSenderIsAnAliasOfTheUserF
         assertThat(canSendFrom().userCanSendFrom(USER, fromGroup)).isFalse();
 
     }
+
+    @Test
+    default void assertAllValidFromAddressesShouldBeEmptyWhenUserHasNoAlias() throws Exception {
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .isEmpty();
+    }
+
+    @Test
+    default void assertAllValidFromAddressesShouldBeEmptyWhenUserHasNoAliasAndAnotherUserHasOne() throws Exception {
+        addAliasMapping(USER_ALIAS, OTHER_USER);
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .isEmpty();
+    }
+
+    @Test
+    default void assertAllValidFromAddressesShouldContainsAnAliasOfTheUser() throws Exception {
+        addAliasMapping(USER_ALIAS, USER);
+
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .contains(USER_ALIAS.asMailAddress());
+    }
+
+    @Test
+    default void assertAllValidFromAddressesFromShouldBeTrueWhenSenderIsAnAliasOfAnAliasOfTheUser() throws Exception {
+        Username userAliasBis = Username.of("aliasbis@" + DOMAIN.asString());
+        addAliasMapping(userAliasBis, USER_ALIAS);
+        addAliasMapping(USER_ALIAS, USER);
+
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .containsExactlyInAnyOrder(USER_ALIAS.asMailAddress(), userAliasBis.asMailAddress());
+    }
+
+    @Test
+    default void assertAllValidFromAddressesShouldContainsAnAliasOfTheDomainUser() throws Exception {
+        Username fromUser = Username.of(USER.getLocalPart() + "@" + OTHER_DOMAIN.asString());
+
+        addDomainMapping(OTHER_DOMAIN, DOMAIN);
+
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .contains(fromUser.asMailAddress());
+    }
+
```

Comment:
Can't we support `userAlias x domainAlias` (1st level of recursion can be obtained through cardinal product)

Reply:
What would be a 2nd level of recursion?

Reply:
To be implemented later

Reply:
this test case is effectively missing, we will add it later